keps/sig-node/4815-dra-partitionable-devices/README.md

```diff
@@ -0,0 +1,3411 @@
+		// to extend the set of attributes and capacities it already defines.
+		//
+		// The main purposes of these mixins is to reduce the memory footprint
+		// of devices since they can reference the mixins provided here rather
+		// than duplicate them.
+		//
+		// Must not have more than 128 entries.
+		//
+		// +optional
+		// +listType=atomic
+		DeviceMixins []DeviceMixin
+}
+
+// DeviceMixin defines a specific device mixin for each device type.
+// Besides the name, exactly one field must be set.
+type DeviceMixin struct {
+		// Name is a unique identifier among all mixins managed by the driver
+		// in the pool. It must be a DNS label.
+		//
+		// +required
+		Name string
+
+		// Basic defines a mixin usable by a basic device.
+		//
+		// +optional
+		// +oneOf=deviceMixinType
+		Basic *BasicDeviceMixin
+}
+
+// BasicDeviceMixin defines a mixin that a basic device can include.
+type BasicDeviceMixin struct {
+		// Includes defines the set of other mixins that this mixin includes.
+		//
+		// The propertes of each included mixin are applied to this mixin in
+		// order (recursively). Conflicting properties from multiple mixins are
+		// taken from the last mixin listed that contains them.
+		//
+		// The maximum number of mixins that can be included is 8.
+		//
+		// +optional
+		Includes []DeviceMixinRef
+
+		// Attributes defines the set of attributes for this mixin.
+		// The name of each attribute must be unique in that set.
+		//
+		// The maximum number of attributes and capacities combined is 32.
+		//
+		// +optional
+		Attributes map[QualifiedName]DeviceAttribute
+
+		// Capacity defines the set of capacities for this mixin.
+		// The name of each capacity must be unique in that set.
+		//
+		// The maximum number of attributes and capacities combined is 32.
+		//
+		// +optional
+		Capacity map[QualifiedName]DeviceCapacity
+}
+
+// Device represents one individual hardware instance that can be selected based
+// on its attributes. Besides the name, exactly one field must be set.
+type Device struct {
+		// Name is unique identifier among all devices managed by
+		// the driver in the pool. It must be a DNS label.
+		//
+		// +required
+		Name string
+
+		// Basic defines one device instance.
+		//
+		// +optional
+		// +oneOf=deviceType
+		Basic *BasicDevice
+}
+
+// BasicDevice defines one device instance.
+type BasicDevice struct {
+		// Includes defines the set of device mixins that this device includes.
+		//
+		// The propertes of each included mixin are applied to this device
+		// in order (recursively). Conflicting properties from multiple mixins
+		// are taken from the last mixin listed that contains them.
+		//
+		// The maximum number of mixins that can be included is 8.
+		//
+		// +optional
+		Includes[]DeviceMixinRef
+
+		// ConsumesCapacityFrom defines the set of devices where any capacity
+		// consumed by this device should be pulled from. This applies recursively.
```

Comment:
This "pulled from" implies that there must be a device that can be pulled from. That device itself gets listed and thus can get allocated.

Suppose we have a flat list of identical devices which have some conflicting capacity requirements. We would need to define a fictional additional device that provides this scarce capacity. To Kubernetes, that device looks like something that it can allocate and a request for a "real" device will match it (same driver name, for example).

This is not insurmountable: the fictional device could have "fictional: true" attribute and the DeviceClass selector could exclude devices where that value is true. I just wanted to call this out. Might be worth adding to the KEP as an example.

Reply:
Yes, this is one limitation of this approach (which I thought of but didn't call out explicitly here).

My original proposal (from many months ago) allowed for a separate place to list "shared" capacities that weren't attached to any particular device. Individual devices could then pull rom this list of shared capacities rather than pulling from another device explicitly.

So in the case of GPUs with MIG, we would have one shared capacity per full GPU from which both full GPUs and MIG devices would pull.

@thockin wasn't a fan of this approach though as he thought it seemed too abstract to have a "bag" of resources just declared in a resource slice but not explicitly tied to any particular physical device. At this point, I'm honestly OK with either approach, it's more about what is the common case, and I think needing to pull from some abstract source is actually the outlier, so it's fine to optimize for the more intuitive case.

Regarding having a "fictional: true attribute". The first version of this KEP had something similar with an explicit field  of `abstract: true`) (not an attribute but a top-level field in the `Device` struct). Having an attribute is just as good (maybe better) though, so long as your `DeviceClass` includes the ability to filter out these types of devices from actually being allocated.

Reply:
A smaller API is better, so I prefer the approach with the attribute. We just need to explain it because it wouldn't be obvious to driver authors how to do this.